core/src/main/scala/org/apache/spark/memory/MemoryManager.scala

```diff
@@ -255,7 +256,17 @@ private[spark] abstract class MemoryManager(
     }
     val size = ByteArrayMethods.nextPowerOf2(maxTungstenMemory / cores / safetyFactor)
     val default = math.min(maxPageSize, math.max(minPageSize, size))
-    conf.get(BUFFER_PAGESIZE).getOrElse(default)
+    val sizeAsBytes = conf.get(BUFFER_PAGESIZE).getOrElse(default)
+    // If we are using G1 GC, it's better to take the LONG_ARRAY_OFFSET into consideration
+    // so that the requested memory size is power of 2 and can be divided by G1 region size
+    // to reduce memory waste within one G1 region
+    if (Utils.isG1GarbageCollector &&
+      tungstenMemoryMode == MemoryMode.ON_HEAP &&
+      sizeAsBytes % (1024 * 1024) == 0) {
+      sizeAsBytes - Platform.LONG_ARRAY_OFFSET
```

Comment:
If I get this right in case of G1GC the best would be if we choose a pageSize where the following holds:

```
  G1HeapRegionSize % (pageSize + Platform.LONG_ARRAY_OFFSET) == 0
```
And when When BUFFER_PAGESIZE is not set we are free to choose it as:
```
pageSize = G1HeapRegionSize - Platform.LONG_ARRAY_OFFSET;
```

And with the above code we just try to calculate G1HeapRegionSize with our own way. But what about accessing this value?
Like the same way used in the [OpenJDK tests](https://www.programcreek.com/java-api-examples/?code=AdoptOpenJDK%2Fopenjdk-jdk8u%2Fopenjdk-jdk8u-master%2Fhotspot%2Ftest%2Fgc%2Farguments%2FTestG1HeapRegionSize.java):

```java
  HotSpotDiagnosticMXBean diagnostic = ManagementFactoryHelper.getDiagnosticMXBean();
  option = diagnostic.getVMOption("G1HeapRegionSize");
```

As I see the OpenJDK test was executed like:
```
run main/othervm -Xmx64m TestG1HeapRegionSize 1048576
```
So `diagnostic.getVMOption("G1HeapRegionSize")` gives back the calculated region size.

Reply:
Yes, it's better to make sure `G1HeapRegionSize % (pageSize + Platform.LONG_ARRAY_OFFSET) == 0`.

But when BUFFER_PAGESIZE is not set, I'm not quite sure if it's reasonable to choose it as `pageSize = G1HeapRegionSize - Platform.LONG_ARRAY_OFFSET`,  which seems a bit different with current logic to get default page size. 
By following the current logic to calculte default page size and then minus  `Platform.LONG_ARRAY_OFFSET` can also make sure `G1HeapRegionSize % (pageSize + Platform.LONG_ARRAY_OFFSET) == 0`

Reply:
>But when BUFFER_PAGESIZE is not set, I'm not quite sure if it's reasonable to choose it as pageSize = >G1HeapRegionSize - Platform.LONG_ARRAY_OFFSET, which seems a bit different with current logic to get default >page size.

How the current logic for the default handles a case where a custom -XX:G1HeapRegionSize is given as extra java options?

Reply:
I see what you mean, you are right

Reply:
@attilapiros Since -XX:G1HeapRegionSize can only be set to N-th power of 2, and the default size calculated here is also N-th power of 2, so it can be guanranteed that  `G1HeapRegionSize % (pageSize + Platform.LONG_ARRAY_OFFSET) == 0`  or  `(pageSize + Platform.LONG_ARRAY_OFFSET) % G1HeapRegionSize == 0`, right?  
Such a change has little effect on the current logic.